thewalrus/quantum/gaussian_checks.py

```diff
@@ -125,24 +130,27 @@ def fidelity(mu1, cov1, mu2, cov2, hbar=2, rtol=1e-05, atol=1e-08):
     if not n0 == n1 == m0 == m1 == l0 == l1:
         raise ValueError("The inputs have incompatible shapes")
 
-    v1 = cov1 / hbar
-    v2 = cov2 / hbar
-    deltar = (mu1 - mu2) / np.sqrt(hbar / 2)
-    n = n0 // 2
-    W = sympmat(n)
-
-    si12 = np.linalg.inv(v1 + v2)
-    vaux = W.T @ si12 @ (0.25 * W + v2 @ W @ v1)
-    p1 = vaux @ W
-    p1 = p1 @ p1
-    p1 = np.identity(2 * n) + 0.25 * np.linalg.inv(p1)
-    if np.allclose(p1, 0, rtol=rtol, atol=atol):
-        p1 = np.zeros_like(p1)
+    # We first convert all the inputs to quantities where hbar = 1
+    sigma1 = cov1 / hbar
+    sigma2 = cov2 / hbar
+    deltar = (mu1 - mu2) / np.sqrt(hbar)
+
+    Omega = sympmat(n0 // 2)  # The symplectic matrix
+
+    Sigma = sigma1 + sigma2
+    Sigma_inv = np.linalg.inv(Sigma)
+    Vaux = Omega.T @ Sigma_inv @ (0.25 * Omega + sigma2 @ Omega @ sigma1)
+    sqrtm_arg = np.identity(n0) + 0.25 * np.linalg.inv(Vaux @ Omega @ Vaux @ Omega)
+    # The sqrtm function has issues with matrices that are close to zero, hence we branch
+    if np.allclose(sqrtm_arg, 0, rtol=rtol, atol=atol):
+        mat_sqrtm = np.zeros_like(sqrtm_arg)
     else:
-        p1 = sqrtm(p1)
-    p1 = 2 * (p1 + np.identity(2 * n))
-    p1 = p1 @ vaux
-    f = np.sqrt(np.linalg.det(si12) * np.linalg.det(p1)) * np.exp(
-        -0.25 * deltar @ si12 @ deltar
+        mat_sqrtm = sqrtm(sqrtm_arg)
+    det_arg = 2 * (mat_sqrtm + np.identity(n0)) @ Vaux
+    f = np.sqrt(np.linalg.det(Sigma_inv) * np.linalg.det(det_arg)) * np.exp(
+        -0.5 * deltar @ Sigma_inv @ deltar
```

Comment:
No transpose on the first `deltar`? It seems like the paper writes it `np.exp(-0.5 * deltar.T @ Sigma_inv @ deltar)`, but maybe that's the same in this case. :thinking:

Reply:
It makes no difference, although it could be added. If you do `a @ b` and `a` and `b` are 1D arrays the numpy is smart enough to know that what you want is mathematically speaking `a^T \cdot b`

Reply:
Fair enough. That's what I suspected, just wanted to clarify. Thanks!